File: code/geometry_cpp/convex_hull.cpp
auto convexHull(vector<ipoint> pts) {
	if (pts.size() <= 1) return pts;
	sort(all(pts));
	decltype(pts) h;
	auto f = [&] (ll s) {
		for (auto p : pts) {
			while ((ll)h.size() >= s + 2 && h.back().cross(p, h[h.size() - 2]) <= 0)
				h.pop_back();
			h.push_back(p);
		}
		h.pop_back();
	};
	f(0);
	reverse(all(pts));
	f(h.size());
	if (h.size() == 2 && h[0] == h[1]) h.pop_back();
	return h;
}

File: code/geometry_py/convex_hull.py
def convexHull(pts):
	if len(pts) <= 1:
		return pts
	pts.sort()
	t, s, h = 0, 0, [0] * (len(pts) + 1)
	for i in range(2):
		for p in pts:
			while t >= s + 2 and cross(h[t - 1], p, h[t - 2]) <= 0:
				t -= 1
			h[t], t = p, t + 1
		s = t = t - 1
		pts.reverse()
	return h[:t - (t == 2 and h[0] == h[1])]
